cirq/study/flatten_expressions.py

```diff
@@ -0,0 +1,368 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Resolves symbolic expressions to unique symbols."""
+
+from typing import (overload, Any, Callable, Dict, Iterator, List, Optional,
+                    Tuple, Union)
+
+import sympy
+
+from cirq import protocols, value
+from cirq.study import resolver, sweeps, sweepable
+
+
+def flatten(val: Any) -> Tuple[Any, 'ExpressionMap']:
+    """Creates a copy of `val` with any symbols or expressions replaced with
+    new symbols.  `val` can be a `Circuit`, `Gate`, `Operation`, or other
+    type.
+
+    Use `ExpressionMap.transform_sweep` or `ExpressionMap.transform_params`
+    after `flatten`.  Use `flatten_with_sweep` or `flatten_with_params` instead
+    of `flatten` as a shortcut.
+
+    Args:
+        val: The value to copy with substituted parameters.
+
+    Example:
+        >>> qubit = cirq.LineQubit(0)
+        >>> a = sympy.Symbol('a')
+        >>> circuit = cirq.Circuit.from_ops(
+        ...     cirq.X(qubit) ** (a/4),
+        ...     cirq.Y(qubit) ** (1-a/2),
+        ... )
+        >>> print(circuit)
+        0: ───X^(a/4)───Y^(1 - a/2)───
+
+        >>> sweep = cirq.Linspace(a, start=0, stop=3, length=4)
+        >>> list(sweep)  #doctest: +NORMALIZE_WHITESPACE
+        [cirq.ParamResolver(OrderedDict([('a', 0.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 1.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 2.0)])),
+         cirq.ParamResolver(OrderedDict([('a', 3.0)]))]
+
+        >>> c_flat, expr_map = cirq.flatten(circuit)
+        >>> print(c_flat)
+        0: ───X^(<a/4>)───Y^(<1 - a/2>)───
+        >>> expr_map
+        cirq.ExpressionMap({a/4: <a/4>, 1 - a/2: <1 - a/2>})
+
+        >>> new_sweep = expr_map.transform_sweep(sweep)
+        >>> list(new_sweep)  #doctest: +NORMALIZE_WHITESPACE
+        [cirq.ParamResolver(OrderedDict([('<a/4>', 0.0), ('<1 - a/2>', 1.0)])),
+         cirq.ParamResolver(OrderedDict([('<a/4>', 0.25), ('<1 - a/2>', 0.5)])),
+         cirq.ParamResolver(OrderedDict([('<a/4>', 0.5), ('<1 - a/2>', 0.0)])),
+         cirq.ParamResolver(OrderedDict([('<a/4>', 0.75), ('<1 - a/2>', -0.5)])\
+)]
```

Comment:
This line was 1 char over 80.  I'm not sure how else to wrap it so that doctest succeeds.